core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1436,17 +1438,15 @@ class ReplicaManager(val config: KafkaConfig,
 
           leaderAndIsrRequest.partitionStates.forEach { partitionState =>
```

Comment:
I wonder if we really have to re-iterate over all the partition states present in the request here. Intuitively, I would have thought that considering only the ones in `partitionStates` would be sufficient.

Reply:
This logic is a bit strange. It sounds like we need are trying to handle the case where we fail to create the log after we have already created the partition. Would it make more sense to handle this in `getOrCreatePartition` if an exception is raised?

Reply:
There's actually another copy of codes doing the same thing in `ReplicaManager.scala : 2097`. I think we can make it a method to avoid code duplication.

Reply:
`getOrCreatePartition` is not even called in this particular case but anyway `createLogIfNotExists` is called later so it would not work.

Reply:
@hachikuji @showuon I have refactored this part. I think that we could remove this check and replace it by making the partition offline when an `KafkaStorageException` occurs. Does it make sense?